### DISCUSSION TITLE
Security whole due to unlimed read/udpate/delete: possibility to disable to read/update/delete multiple items

### DISCUSSION THREAD

**Discussion by dsolodkyicf** [directus/directus#24075] [Feature Requests]
### Summary

The currrent permssions system makes no difference between the action of the user, when he wants to read/update/delete a complete collection or just a single item. It allows to scrape the database, update all items with bad data or even to remove all items from the collection. 

### Motivation

1) Currently by exposing the READ access, I allow the user to scrape my complete database -> securoty whole
2) By exposing UPDATE permission, I allow the user change every item if the collection, he has access to (by applying "item permission" with a single REST request.

So that a HUGE security whole, which mush be fixed.

### Detailed Design

Add a checkbox on Item Permission: for Read, Update, Create, Delete, Share: "Alow multple items"

### Must Have:
- It shall be possible to limit the READ pemission in way that user can read only a single item via REST/GraphQL
- It shall not be possible to delete all records the user has accesss to via a single rest response.

### Drawbacks

none

### Alternatives

Just a security whole. I am wondering why it has not be done yet.

### Adoption Strategy

If it is a checkbox/boolean field, it shall not be the breaking changes

**Comment by rijkvanzanten** [comment]
This is already tracked over here as well: https://github.com/directus/directus/discussions/11464 🙂 I'll close yours so we don't end up spreading the votes across two different posts, but I'll make sure to link it so we don't lose valuable context. Thanks!

**Reply by dsolodkyicf** [reply]
There is quite some difference betwene just readOne and updateOne/deleteOne.

**Reply by rijkvanzanten** [reply]
Your two requirements are already possible; it's called "Item Permissions". You can setup any filter rule that'll be checked before the action is allowed. That includes setting "ID has to be 1" to limit it down to just a single item by ID 👍🏻

**Reply by dsolodkyicf** [reply]
Id has to be "1" means that only item with very specific id=1, can be modified. It is really different from the approach when user can change or delete any single item with the known id, but single item at the time(per single request). Let's imagine you have the id as UUID, then user can only change the item if he knows uuid, but in the current approach, he can change/delete every item he might have access to with a single request.

The read permission is even a bigger security hole: if you can read some items, you can read all of them in one request. So if you have, let's say articles, with valuable content, you need the public users to have read access to them. So any public user can scrape all of the articles with complete content with a single GET request without filters.

So I am looking for the way to enforce the (public) user to provide an id when he does a request, and i do not see any except by creating extensions and hiding the standard directus api inside the internal network, as they are not secure enough.

**Reply by rijkvanzanten** [reply]
> Let's imagine you have the id as UUID, then user can only change the item if he knows uuid, but in the current approach, he can change/delete every item he might have access to with a single request.

That's wrong. If you configure your permissions with an Item Permissions rule of "ID = 1" they can only ever interact with item where that item permission rule matches 👍🏻 Same goes for reads, if you configure your item permissions for the read action with "status = 'published'", they can now only read the items where the status field equals "published".

The whole "I only want to allow them to be able to make a request to one item at a time" part is the feature request I linked earlier 🙂

**Reply by dsolodkyicf** [reply]
What exactly is wrong?

Don't you see a difference between Updating a specific, single item vs Updating all items user has access to (in accordance to permission rule)?

Let's say a user is content editor. So he can edit (or delete) all articles. Why is it ok to provide him a possibility to Update or Delete all articles at once with a single rest request? No, you do want the user to only be able to do it one by one.

**Reply by manuelwehrli** [reply]
Why should a user not be able to update or delete multiple items at once if he has permissions and knows the primary keys (which is necessary for such a request)? This is an essential feature of the directus app so the api must support it, better than firing a huge amount of single delete requests from the frontend. I mean, if one wants to be a bad actor, one can delete one by one with a curl script, where only some rate limiting could stop such behavior, which is only configurable on a global request level, not specifically for one api action. If you want to protect your data from somebody going wild, you have to enable revisions for updates and configure the archiving mechanism and delete those items later as admin or some kind of supervisor role, or automatically with a flow after some time. The only problem (which is addressed by that other feature request) is actually that the read permission enables listing of all items and makes it easy to mine a lot of data with one request that should only be available partially in the context of other data. A common example is the files collection where we often need broad (public) read access so assets can be fetched in different contexts, but we still do not want to expose all stored assets as a list. Or do you still see a use case that is not covered after all this would be implemented or configured by you?

**Reply by dsolodkyicf** [reply]
Delete does *not* require ids to be provided. The filters are supported in delete. So the user can delete all items without knowing their ids. I don't know how shall it be ok at all..

**Reply by manuelwehrli** [reply]
I saw it now in the docs, you are right, deleteItems works with ids or a query. What speaks against strictly configuring the archiving mechanism? Even if deleteItems would only work with explicit ids or only deleteItem would exist, one can still obtain primary keys of many items in the one or other way. Totally disabling the listing is not always an option even if possible, most of the time, we need some way to get all items a user has read permission for to display a list for further interaction. The only way to really avoid mass deletion is not giving delete permission at all as I described.

**Reply by dsolodkyicf** [reply]
Once again: delete item by id is ok. Knowing id and having delete permission allowes you to do so. But allowing deleting ALL items you potentially has access to with a single request? You really don't want to provide this functionality to the limited user. At least you need to have an option to disable it. This is basically what I am asking for: please allow me to disable deleteItems for a specific user group. Then this ticket + possibility to disable getItems (while keeping getItem enabled: so another ticket, mentioned above) will allow me to implement additional obstacles for the user to not get (and remove) a complete (sub-)set of collection via additional business logic on backend and front end. It just shall not be easy for the user to make bad things, even if he is allowed to do so.
